sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamIOPushDownRule.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilder;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.util.Pair;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+
+public class BeamIOPushDownRule extends RelOptRule {
+  // ~ Static fields/initializers ---------------------------------------------
+
+  public static final BeamIOPushDownRule INSTANCE =
+      new BeamIOPushDownRule(RelFactories.LOGICAL_BUILDER);
+
+  // ~ Constructors -----------------------------------------------------------
+
+  public BeamIOPushDownRule(RelBuilderFactory relBuilderFactory) {
+    super(operand(Calc.class, operand(BeamIOSourceRel.class, any())), relBuilderFactory, null);
+  }
+
+  // ~ Methods ----------------------------------------------------------------
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final Calc calc = call.rel(0);
+    final BeamIOSourceRel ioSourceRel = call.rel(1);
+    final BeamSqlTable beamSqlTable = ioSourceRel.getBeamSqlTable();
+    final RexProgram program = calc.getProgram();
+    final Pair<ImmutableList<RexNode>, ImmutableList<RexNode>> projectFilter = program.split();
+    final RelDataType calcInputRowType = program.getInputRowType();
+    RelBuilder relBuilder = call.builder();
+
+    if (!beamSqlTable.supportsProjects()) {
+      return;
+    }
+
+    // Nested rows are not supported at the moment
+    for (RelDataTypeField field : ioSourceRel.getRowType().getFieldList()) {
+      if (field.getType() instanceof RelRecordType) {
+        return;
+      }
+    }
+
+    // Find all input refs used by projects
+    Set<String> usedFields = new HashSet<>();
+    for (RexNode project : projectFilter.left) {
+      findUtilizedInputRefs(calcInputRowType, project, usedFields);
+    }
+
+    // Check if the calc can be dropped
+    if (isProjectRenameOnlyProgram(program)) {
+      call.transformTo(ioSourceRel.copy(calc.getRowType(), new ArrayList<>(usedFields)));
+      return;
+    }
+
+    // Find all input refs used by filters
+    for (RexNode filter : projectFilter.right) {
+      findUtilizedInputRefs(calcInputRowType, filter, usedFields);
+    }
+
+    // TODO: will need to be updated for predicate push-down
+    if (usedFields.size()
+        == ioSourceRel.getRowType().getFieldCount()) { // Already most optimal case
+      return;
+    }
+
+    FieldAccessDescriptor resolved =
+        FieldAccessDescriptor.withFieldNames(usedFields).resolve(beamSqlTable.getSchema());
+    Schema newSchema =
+        SelectHelpers.getOutputSchema(ioSourceRel.getBeamSqlTable().getSchema(), resolved);
+    RelDataType calcInputType =
+        CalciteUtils.toCalciteRowType(newSchema, ioSourceRel.getCluster().getTypeFactory());
+
+    BeamIOSourceRel newIoSourceRel = ioSourceRel.copy(calcInputType, new ArrayList<>(usedFields));
```

Comment:
Just converting to an `ArrayList` is a little scary to me. Is this list guaranteed to be in the same order as the output row? (I'm guessing there are IOs that would expect it to be.)

Reply:
Changed to use `newSchema.getFieldNames()` instead, which should produce the list of names in the same order they are present in the original schema.